src/state/shared/assets.ts

```diff
@@ -0,0 +1,36 @@
+import { makeAutoObservable, runInAction, when } from 'mobx';
+import { ViewService } from '@penumbra-zone/protobuf';
+import { Metadata } from '@penumbra-zone/protobuf/penumbra/core/asset/v1/asset_pb';
+import { getDenomMetadata } from '@penumbra-zone/getters/assets-response';
+import { penumbra } from '@/utils/penumbra';
+import { connectionStore } from '@/state/connection';
+
+class AssetsState {
+  /** If true, ignore all other state values */
+  loading = false;
+  /** The error is set in case of a request failure */
+  error?: string;
+  assets: Metadata[] = [];
+
+  constructor() {
+    makeAutoObservable(this);
+
+    when(
+      () => connectionStore.connected,
+      () => void this.setup(),
+    );
+  }
+
+  async setup() {
+    try {
+      runInAction(() => this.loading = true);
+      const responses = await Array.fromAsync(penumbra.service(ViewService).assets({}));
+      this.assets = responses.map(getDenomMetadata);
+      this.loading = false;
```

Comment:
issue: Async mobx actions are a bit tricky. It looks like you are doing `runInAction()` only for one of the fields. My suggestion is to create a `setLoading()` and `setAssets()` and `setError()`. Otherwise, you'll get a warning from mobx in the console for async functions.

Reply:
Realized the loading state was unnecessary. As for the rest, applied the setters as you proposed